p2p/src/peer_manager/address_groups.rs

```diff
@@ -0,0 +1,99 @@
+// Copyright (c) 2023 RBB S.r.l
+// opensource@mintlayer.org
+// SPDX-License-Identifier: MIT
+// Licensed under the MIT License;
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://github.com/mintlayer/mintlayer-core/blob/master/LICENSE
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::net::{Ipv4Addr, Ipv6Addr};
+
+use crate::types::peer_address::PeerAddress;
+
+use super::global_ip::IsGlobalIp;
+
+// IPv4 addresses /16 groups
+const IPV4_GROUP_BYTES: usize = 2;
+// IPv4 addresses /32 groups
+const IPV6_GROUP_BYTES: usize = 4;
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub enum AddressGroup {
+    Local,
+    Private,
+    PublicV4([u8; IPV4_GROUP_BYTES]),
+    PublicV6([u8; IPV6_GROUP_BYTES]),
+}
+
+/// Get the canonical identifier of the network group for address.
+///
+/// The groups are assigned in a way where it should be costly for an attacker to
+/// obtain addresses with many different group identifiers, even if it is cheap
+/// to obtain addresses with the same identifier.
+///
+/// See `NetGroupManager::GetGroup` in Bitcoin Core for a reference.
+pub fn get_address_group(address: &PeerAddress) -> AddressGroup {
```

Comment:
This looks like should be an associated function in `AddressGroup`, with a name like `from_peer_address()` or something.